### PR TITLE
Multi_ip allocation

### PR DESCRIPTION
**Pull request by StrongMonkey**

### CODE REVIEW

**Comment by StrongMonkey** [comment]
https://github.com/rancher/rancher/issues/7034

**Comment by StrongMonkey** [comment]
https://github.com/rancher/rancher/issues/3703

File: code/iaas/simple-allocator/src/main/java/io/cattle/platform/simple/allocator/dao/impl/SimpleAllocatorDaoImpl.java [hunk 1]
```diff
@@ -199,5 +212,48 @@ public ObjectManager getObjectManager() {
     public void setObjectManager(ObjectManager objectManager) {
         this.objectManager = objectManager;
     }
+    
+    
+    @Override
+    public ResourceRequest populateResourceRequestFromInstance(Instance instance, String resourceType, String poolType) {
```
Comment (cjellick) [inline code/iaas/simple-allocator/src/main/java/io/cattle/platform/simple/allocator/dao/impl/SimpleAllocatorDaoImpl.java]: If you want to encapsulate all this logic in a method, that is fine, but a Dao method is not the right place. Dao methods should deal solely with retrieving data from the DB and have very little knowledge of the system/business logic outside of that.

You could move a method like this to simpleAllocator

File: code/iaas/allocator/src/main/java/io/cattle/platform/allocator/dao/impl/AllocatorDaoImpl.java [hunk 2]
```diff
@@ -486,4 +494,57 @@ public boolean hostHasContainerLabel(long hostId, String labelKey,
...
+     *                              ]
+     *                          }
+     * }
+     *      
+     * Then update 
+     * Port Table. Binding address field in port table and public port field in port table if public port is allocated by external scheduler (for agent)  
+     */
+    @SuppressWarnings("unchecked")
+    @Override
+    public void updateInstancePorts(Map<String, Object> data) {
```
Comment (cjellick) [inline code/iaas/allocator/src/main/java/io/cattle/platform/allocator/dao/impl/AllocatorDaoImpl.java]: To refactor, you can add a ports field to AlllocationAttempt. You can update it upon receiving the reserve response, but then move the logic for saving that to the instance into io.cattle.platform.allocator.dao.AllocatorDao.recordCandidate(AllocationAttempt, AllocationCandidate)

Comment (StrongMonkey) [inline code/iaas/allocator/src/main/java/io/cattle/platform/allocator/dao/impl/AllocatorDaoImpl.java]: roger

File: code/iaas/simple-allocator/src/main/java/io/cattle/platform/simple/allocator/dao/impl/SimpleAllocatorDaoImpl.java (hunk 1 above)
Comment (StrongMonkey) [inline code/iaas/simple-allocator/src/main/java/io/cattle/platform/simple/allocator/dao/impl/SimpleAllocatorDaoImpl.java]: yeah, will do

**Comment by StrongMonkey** [comment]
@cjellick  i made the changes you requested.  https://github.com/rancher/validation-tests/pull/274 added the validation test.

**Comment by StrongMonkey** [comment]
pr rebased.

File: code/iaas/allocator/src/main/java/io/cattle/platform/allocator/service/AllocationAttempt.java [hunk 3]
```diff
@@ -136,4 +138,12 @@ public void setPoolIds(Map<Long, Set<Long>> poolIds) {
         this.poolIds = poolIds;
     }
 
+    public Map<String, Object> getAllocatedIPs() {
+        return allocatedIPs;
```
Comment (cjellick) [inline code/iaas/allocator/src/main/java/io/cattle/platform/allocator/service/AllocationAttempt.java]: Remind that based on our conversation regarding the bug in the scheduler, you need to make this a list.

File: code/iaas/simple-allocator/src/main/java/io/cattle/platform/simple/allocator/SimpleAllocator.java [hunk 4]
```diff
@@ -64,7 +73,7 @@
 
     @Inject
     AgentLocator agentLocator;
-
+
```
Comment (cjellick) [inline code/iaas/simple-allocator/src/main/java/io/cattle/platform/simple/allocator/SimpleAllocator.java]: Nooooo!

Comment (StrongMonkey) [inline code/iaas/simple-allocator/src/main/java/io/cattle/platform/simple/allocator/SimpleAllocator.java]: `go fmt -s`

**Comment by StrongMonkey** [comment]
no metadata changes, fixed the bug you mentioned

File: code/iaas/simple-allocator/src/main/java/io/cattle/platform/simple/allocator/PortBindingResourceRequest.java [hunk 5]
```diff
@@ -0,0 +1,46 @@
+package io.cattle.platform.simple.allocator;
+
+import io.cattle.platform.core.util.PortSpec;
+
+import java.util.List;
+
+public class PortBindingResourceRequest implements ResourceRequest {
+    private String instanceId;
+    private String resource;
+    private List<PortSpec> PortRequests;
```
Comment (cjellick) [inline code/iaas/simple-allocator/src/main/java/io/cattle/platform/simple/allocator/PortBindingResourceRequest.java]: Whoops capitalized! Too much golang to java context switching

**Pull request closed without merging**